dremio_toolkit/rebuild_metadata.py

```diff
@@ -0,0 +1,129 @@
+        print("datasource argument has not been specified. Metadata for all physical datasets in all data sources will be rebuilt.")
+    if int(parsed_args.concurrency) > 4:
+        print("Specifying concurrency higher than 4 without setting dremio.iceberg.enabled to True may impact Dremio performance and potentially bring Dremio cluster down.")
+    return parsed_args
+
+
+def rebuild_metadata(dremio_environment_url, user, password, datasource, concurrency, report_filename, log_level, log_filename):
+    logger = Logger(level=log_level, log_file=log_filename)
+    logger.new_process_status(1, 'Retrieving list of PDS for rebuilding metadata ...')
+    env_api = EnvApi(dremio_environment_url, user, password, logger)
+    pds_list = get_pds_list(env_api, datasource)
+    if pds_list is None:
+        logger.fatal("Unable to retrieve list of PDS. ")
+    logger.new_process_status(len(pds_list), 'Rebuilding metadata. ')
+    base_threads_count = threading.activeCount()
+    threads = []
+    for pds in reversed(pds_list):
+        # Wait for a thread to become available to not exceed specified concurrency
+        while threading.activeCount() >= base_threads_count + concurrency:
+            time.sleep(3)
+        new_thread = RebuildMetadataThread(env_api, pds, logger)
+        new_thread.start()
+        threads.append(new_thread)
+
+    # Wait for all threads to finish
+    job_statuses = []
+    for thread in threads:
+        if thread.is_alive():
+            thread.join()
+        if thread.get_status():
+            job_statuses.append({'pds': thread.get_pds_path(),
+                                 'job_id': thread.get_job_id(),
+                                 'pds_rebuild_status': 'SUCCESS' if thread.get_status() else 'FAILED',
+                                 'forget_job_info': thread.get_forget_job_info(),
+                                 'refresh_job_info': thread.get_refresh_job_info()})
+
+    # Produce execution report
+    if report_filename:
+        if os.path.isfile(report_filename):
+            os.remove(report_filename)
+        with open(report_filename, "w", encoding="utf-8") as f:
+            json.dump(job_statuses, f, indent=4, sort_keys=True)
+
+    logger.finish_process_status_reporting()
+    if logger.get_error_count() > 0:
+        exit(Utils.NON_FATAL_EXIT_CODE)
+
+
+def get_pds_list(env_api, datasource) -> list:
+    sql = 'SELECT TABLE_SCHEMA, TABLE_NAME FROM INFORMATION_SCHEMA."TABLES" WHERE TABLE_TYPE = \'TABLE\''
+    if datasource:
+        sql += ' AND ( \'' + datasource.lower() + '\' = LOWER(TABLE_SCHEMA) OR ' +\
+                       'POSITION(\'' + datasource.lower() + '.\' IN LOWER(TABLE_SCHEMA)) = 1)'
+    status, jobid, job_result = env_api.execute_sql(sql)
+    if not status:
+        return None
+    num_rows = int(job_result['rowCount'])
+    # Page through the results, 100 rows per page
+    limit = 100
+    pds_list = []
+    for i in range(0, int(num_rows / limit) + 1):
```

Comment:
range by default starts from 0
```suggestion
    for i in range(int(num_rows / limit) + 1):
```

Reply:
Done